### ISSUE TITLE
exposing SingleColumnTransformer and RejectColumn

### ISSUE THREAD

**Issue by jeromedockes** [skrub-data/skrub#1730]
### Problem Description

the ApplyToCols has been made public. it becomes even more useful if user-defined transformers can make full use of its functionality, by handling a single Series (makes code easier to write) and by possibly rejecting columns with a RejectColumn exception.

For example suppose I want to write a transformer that parses zip codes and converts to a latitude and longitude. As it processes a single column at a time it's a bit simpler if I can inherit from SingleColumnTransformer and deal with Series rather than DataFrames in fit() and transform(). Also it is useful to just raise a RejectColumn exception to signal that the column does not contain zip codes after all, in which case the ApplyToCols will leave it unchanged. Auto-detecting the zip code columns before transforming them would be much more annoying to do otherwise, and would basically force me to write a full transformer (meaning I wouldn't be able to use ApplyToCols for this)

### Feature Description

SingleColumnTransformer and RejectColumn would be added to the public API, possibly with improvements or additions to the documentation & examples, without any change to the functionality

Both objects should be added to a new `skrub.core` namespace, which includes more "dev-oriented" objects that should not clutter the "general use case" namespace.

**Comment by rcap107** [comment]
I agree that this would be useful, and I think it would also simplify docstrings in other places, letting us refer to the public docstring of SingleColumnTransformer and RejectColumn directly. 

I also think this would be targeted mostly at advanced users, and it should be reflected in the documentation. 

In any case, this should be done after #1729. Both issues may be addressed at the same time, but given that this issue involves some work on the documentation, it might be better to split them.

**Comment by GaelVaroquaux** [comment]
I'm fine with the suggestion of adding SingleColumnTransformer and RejectColumn to the public API.

My question would however be: should they be exposed in skrub's namespace, or in another namespace, such as "skrub.api"?
I'm asking to avoid polluting the tab-completion on the main namespace with things that are not useful aside for advanced users.

**Comment by rcap107** [comment]
> My question would however be: should they be exposed in skrub's namespace, or in another namespace, such as "skrub.api"? I'm asking to avoid polluting the tab-completion on the main namespace with things that are not useful aside for advanced users.

I think that makes sense. I think the use case is advanced enough that people that are interested in it shouldn't have issues with one additional level of imports.

However, my follow-up question to this is: should we reconsider importing everything at once in skrub, and maybe split some of the lower level objects into specific namespaces? Like "joining", "encoding" etc

TableVectorizer, tabular_pipeline, TableReport would all stay in the main namespace

**Comment by GaelVaroquaux** [comment]
> However, my follow-up question to this is: should we reconsider importing everything at once in skrub, and maybe split some of the lower level objects into specific namespaces? Like "joining", "encoding" etc

I would rather keep as much as possible in the main namespace (we don't have everything: we have selectors). Experience (eg with sklearn) shows that remembering where things are is a major cognitive overhead for users.

**Comment by jeromedockes** [comment]
I also quite like having everything in the top-level module. I think having some things in a separate module is nice when they are a bit less frequently used, _and_ it is immediately obvious where they would go so it's easy to remember or guess: the `selectors` is a good example of that.

I agree singlecolumntransformer and rejectcolumn are more advanced / less frequently used, however maybe we can find a grouping that is a bit more intuitive than 'api' otherwise I'm afraid it will be hard for users to get a feel of when they should look for something in skrub and when they should look for it in api

**Comment by jeromedockes** [comment]
regarding tab completion, WDYT of defining a `skrub.__dir__()` in which we list the most frequently-used names? At the moment, there are already too many for tab-completion to be very useful if you don't know what you're looking for well enough to type a prefix (there are 52 names, some of which are never used directly by user code like SkrubLearner). the other names would still be in the skrub module, just wouldn't show up in tab completion. and we can curate this list and move things in and out of it without breaking backwards compatibility.

I would at least exclude from it names that are not used by users like SkrubLearner, which is there mostly so that it shows up in the docs

**Comment by rcap107** [comment]
> I agree singlecolumntransformer and rejectcolumn are more advanced / less frequently used, however maybe we can find a grouping that is a bit more intuitive than 'api' otherwise I'm afraid it will be hard for users to get a feel of when they should look for something in skrub and when they should look for it in api

Maybe "base", like in scikit-learn? Given both objects are used to implement a lot of the other transformers

**Comment by rcap107** [comment]
> regarding tab completion, WDYT of defining a `skrub.__dir__()` in which we list the most frequently-used names? At the moment, there are already too many for tab-completion to be very useful if you don't know what you're looking for well enough to type a prefix (there are 52 names, some of which are never used directly by user code like SkrubLearner). the other names would still be in the skrub module, just wouldn't show up in tab completion. and we can curate this list and move things in and out of it without breaking backwards compatibility.

I really like this idea, I did not know of `__dir__()` but it would probably help a lot. I also think the skrub namespace is getting cluttered.

**Comment by GaelVaroquaux** [comment]
> I agree singlecolumntransformer and rejectcolumn are more advanced / less frequently used, however maybe we can find a grouping that is a bit more intuitive than 'api'

I'm happy to have another name

**Comment by GaelVaroquaux** [comment]
> regarding tab completion, WDYT of defining a skrub.__dir__() in which we list the most frequently-used names? 

I think that it is really strange if things that are legitimately useful are not in __dir__.

> At the moment, there are already too many for tab-completion to be very useful if you don't know what you're looking for well enough to type a prefix (there are 52 names, some of which are never used directly by user code like SkrubLearner).

I would rather move these less useful symbol in another namespace (and we need to find a good name)

**Comment by jeromedockes** [comment]
for me the difficulty is that "useful" is subjective and continuous so it might be hard for users to guess (and for us to decide) what goes in skrub and what goes in the sub module.

but maybe we can start by listing the names we don't want in the top-level and then it will be easier to see what they have in common & come up with a name.

one advantage of modules like "encoders" and "joiners" as suggested by @rcap107  is that they are relatively well-defined groups. but that probably doesn't compensate the convenience of having just 1 thing to import in most cases and being able to write for example `skrub.StringEncoder`

**Comment by GaelVaroquaux** [comment]
> for me the difficulty is that "useful" is subjective and continuous

But a lot of user design is exactly this. And a lot of it is decide on what to put attention, to spare attentional load for the users.

> but maybe we can start by listing the names we don't want in the top-level and then it will be easier to see what they have in common & come up with a name.

Yes

> one advantage of modules like "encoders" and "joiners" as suggested by @rcap107 is that they are relatively well-defined groups. but that probably doesn't compensate the convenience of having just 1 thing to import in most cases and being able to write for example skrub.StringEncoder

Yes, I'd rather not go this way.

Actually, I would love us to start more systematically using "import skrub as skb" in our example (seems like a "good first issue"). What do people think? It makes it easier to copy-paste code (and coding is all about copy-pasting code from Internet)

**Comment by rcap107** [comment]
> Actually, I would love us to start more systematically using "import skrub as skb" in our example (seems like a "good first issue"). What do people think? It makes it easier to copy-paste code (and coding is all about copy-pasting code from Internet)

It might also be useful to get users more familiar with the `.skb` namespace for the data ops

**Comment by rcap107** [comment]
Looking at what tab completing skrub does, I noticed a couple of things
- I just found out that `compute_ngram_distance` is a public function, I've never seen it before.
- For some reason, `Drop` (not `DropCols`) is a public class, when it clearly shouldn't be. This should be fixed. 
- There are some entries (`DataOp`, `SkrubLearner`, `ParamSearch`) that are not supposed to be used directly by the user.
- In general, I think that mixing up "regular" skrub objects (TableReport, TableVectorizer etc) and Data Ops-specific objects (like the choose functions) adds a lot of clutter.

I think the last point is the most important. If a user isn't using the Data Ops at all (which likely the case for the majority of skrub users anyway), then a good fraction of what's in the `skrub` namespace is noise. 

However, I'm not convinced that we should move everything Data Op to a different namespace because it would likely cause confusion with the already existing `.skb` namespace. 

Aside from that, I don't see anything specific that I would move to a more specialized namespace.

**Comment by GaelVaroquaux** [comment]
>   • I just found out that compute_ngram_distance is a public function, I've never seen it before.
>   • For some reason, Drop (not DropCols) is a public class, when it clearly shouldn't be. This should be fixed.
>   • There are some entries (DataOp, SkrubLearner, ParamSearch) that are not supposed to be used directly by the user.
>   • In general, I think that mixing up "regular" skrub objects (TableReport, TableVectorizer etc) and Data Ops-specific objects (like the choose functions) adds a lot of clutter.

> I think the last point is the most important. If a user isn't using the Data Ops at all (which likely the case for the majority of skrub users anyway), then a good fraction of what's in the skrub namespace is noise.

I would actually be inclined in fixing the 3 first points and not the last one.

> However, I'm not convinced that we should move everything Data Op to a different namespace because it would likely cause confusion with the already existing .skb namespace.

We could introduce a super short namespace for data ops. I wonder about "do", short for DataOps, but also a verb in English?


```
import skrub as skb
products = skb.do.var("products", products_df)
aggregated_products = products.groupby("basket_ID").agg(
          skb.do.choose_from(["mean", "sum"])
)
```


We would have in it things like:

* do.var
* do.y
* do.X
* do.choose_***
* do.optional
* do.eval_mode
* do.as_data_op

I have the feel that it works quite well in terms of readable and could improve the usability by teasing out what's Data Ops and what is not.

My recommendation:
1. Address the three first points
2. Sleep on the last point, and discuss it more broadly

**Comment by rcap107** [comment]
> My recommendation:
> 1. Address the three first points
> 2. Sleep on the last point, and discuss it more broadly

About the first part: how should we address the points exactly? Is that meant to be done with __dir__, by removing `DataOp` etc? 

About `compute_ngram_distance`, should we set it as private? 

About the ".do" namespace, I was thinking the same. Would that replace the `.skb` namespace, or complement it?

**Comment by GaelVaroquaux** [comment]
> About the first part: how should we address the points exactly? Is that meant to be done with dir, by removing DataOp etc?

Moving them out to another namespace.

> About compute_ngram_distance, should we set it as private?

I don't know. Either this or moving it to another namespace.

> About the ".do" namespace, I was thinking the same. Would that replace the .skb namespace, or complement it?

Hum, interesting. I fear that rename the ".skb" namespace attached to object to ".do" will increase the likehood of name collisions, and thus I would frown from it.

**Comment by jeromedockes** [comment]
> For some reason, Drop (not DropCols) is a public class, when it clearly shouldn't be. This should be fixed.

they are both public, and both are useful. Drop is the single-column transformer that just drops  the column it is applied to. It is public because it is actually used in public attributes of the TableVectorizer when the user passes "drop" as one of the encoders. But it also becomes useful in its own right now that we have the ApplyToCols and the data ops -- for example `.skb.apply(skrub.choose_from([StringEncoder(), Drop()]))`

**Comment by jeromedockes** [comment]
For names that are documented but are never used directly by the user like DataOp, I still think it is reasonable to exclude them from `__dir__()`. According to the documentation for `dir` 

> it is supplied primarily as a convenience for use at an interactive prompt, it tries to supply an interesting set of names more than it tries to supply a rigorously or consistently defined set of names

so DataOp for example is not really "interesting" especially at an interactive prompt. I agree that we should not exclude any names that are ever used by users.

**Comment by jeromedockes** [comment]
about `skb.do` I'm not strictly opposed but it is starting to look a bit noisy & cryptic for someone who lands in the middle of an example and is not familiar with the aliases and shortened names we chose. At this point I might prefer to name it `data_ops` and let users alias it or import some names directly like

```python
from skrub import data_ops
from skrub.data_ops import choose_from

data_ops.X().skb.apply(choose_from(...))
```

`skb.do.X` looks a bit like my 3-year-old got a hold of my keyboard (again!)

it is always possible to 

```python
from skrub import data_ops as do
```

However I'm not sure this is better than the convenience of `skrub.X()`, `skrub.choose_from(...)`

**Comment by jeromedockes** [comment]
> About compute_ngram_distance, should we set it as private?

I'm in favor of making it private. It is too low-level and does not do enough to be useful for users IMO. I has just 2 lines of code to call TfidfVectorizer then pdist, and leaves it up to the caller to make sure input strings are unique and actually do something of the resulting distances, so it really looks like a small private helper to me.

**Comment by rcap107** [comment]
> > For some reason, Drop (not DropCols) is a public class, when it clearly shouldn't be. This should be fixed.
> 
> they are both public, and both are useful. Drop is the single-column transformer that just drops the column it is applied to. It is public because it is actually used in public attributes of the TableVectorizer when the user passes "drop" as one of the encoders. But it also becomes useful in its own right now that we have the ApplyToCols and the data ops -- for example `.skb.apply(skrub.choose_from([StringEncoder(), Drop()]))`

If that is the intended way of using `Drop`, then it makes sense to keep it, but if that's the case it needs a proper docstring because at the moment it's not even in the API docs.

**Comment by rcap107** [comment]
> it is always possible to
> 
> from skrub import data_ops as do
> However I'm not sure this is better than the convenience of `skrub.X()`, `skrub.choose_from(...)`

I quite like this solution, I think I prefer it to having the data ops in the main namespace because they're quite separate from the "regular" skrub operations, and once they're imported as `do` it would be possible to simply write `do.choose_from(...)`.

About skrub -> skb, this looks weird to me:

```python
import skrub as skb

skb.X().skb.apply(...)
```

Here the first `skb` is the main namespace and the second is the dataops namespace and I think that would be very confusing to the users.

**Comment by jeromedockes** [comment]
> Here the first skb is the main namespace and the second is the dataops namespace and I think that would be very confusing to the users.

I also prefer simply `import skrub` than `import skrub as skb`

**Comment by GaelVaroquaux** [comment]
> import skrub as skb

> skb.X().skb.apply(...)

> Here the first skb is the main namespace and the second is the dataops namespace and I think that would be very confusing to the users.

I think that this is a fair point

**Comment by rcap107** [comment]
From today's IRL meeting on skrub: 
- Anything Data Ops should be moved to a `skrub.data_ops` namespace, which should be imported as `import skrub.data_ops as do`. The migration should be done over time (to be discussed how). 
- SingleColumnTransformer and RejectColumns should go in the new namespace `skrub.core` 
- The other issues mentioned above can be tackled in separate, smaller PRs

**Comment by GaelVaroquaux** [comment]
>   • Anything Data Ops should be moved to a skrub.data_ops namespace, which should be imported as import skrub.data_ops as do.

I would do "skrub.do" and not "skrub.data_ops", not counting for users to do the import as

**Comment by rcap107** [comment]
> • Anything Data Ops should be moved to a skrub.data_ops namespace, which should be imported as import skrub.data_ops as do.
> I would do "skrub.do" and not "skrub.data_ops", not counting for users to do the import as

Works for me

**Comment by jeromedockes** [comment]
I don't think it helps users to give the module a meaningless name. Only we can guess that "do" stands for "data ops", and we don't usually give those kinds of names (in skrub or other packages) -- we don't have `skrub.s` for selectors or `skrub.jn` for all joiners. It's already a rather difficult feature to learn and I'm not sure hiding it behind a mysterious name will help

**Comment by jeromedockes** [comment]
Regarding highlighting / directing attention, here are the ones I would de-prioritize:

- All the joiners: use data ops instead. The joiners are unusable in many cases because:
  - they freeze the auxiliary table during training. so for example if we join baskets on the products they contain, all baskets will be empty at prediction time (only the training baskets' contents are available). The airports example in the gallery shows joining weather data to flights with a shuffle split; in a real scenario it would not have any weather info for new flights.
  - they dump the dataset in the pipeline's pickle, making it very large.
  - they cannot transform the aux table besides simple aggregations so lose information in non-numeric columns.
  - agg target is a bit similar to targetencoder but misses cross-fitting or some kind of regularization.
- deduplicate: I haven't seen it work on any real data
- to_datetime: can be useful but more niche
- MinHashEncoder: use StringEncoder or TextEncoder instead
- GapEncoder: can be useful but in most cases use StringEncoder or TextEncoder instead

**Comment by GaelVaroquaux** [comment]
> we don't have skrub.s for selectors or

I think that we should have sel :)

> I'm not sure hiding it behind a mysterious name will help

People don't have to know what the name stands for this to be useful. It's a separation namespace. 

People will mostly learn / pick up data ops by copy-pasting from internet, so I think that it's fine

**Comment by jeromedockes** [comment]
>> we don't have skrub.s for selectors or
> 
> I think that we should have sel :)
> 
>> I'm not sure hiding it behind a mysterious name will help
> 
> People don't have to know what the name stands for this to be useful. It's a separation namespace. 
> 

it's easier to remember things that make sense. virtually none of the widely-used modules in the standard library or in important packages like numpy, scipy, polars etc. or other non-scientific packages have names like that, so I think we should do like others  and have  a meaningful name, and optionally give short aliases when importing. IMO adding 'skrub.sel' or 'skrub.do' would be a regressions (in addition to being breaking changes)

**Comment by rcap107** [comment]
Going back to the original issue, the PR that addresses this should include a section in the user guide that defines a custom transformer and applies it to a dataframe using ApplyToCols. That would be the intended way of using a udf with Apply*

Something like
```python
import pandas as pd
import skrub.selectors as s
from skrub import DropCols

df = pd.DataFrame({"a": [None, None, None, 1], "b": [1,2,3,4]})
class transformer(SingleColumnTransformer):
    def fit_transform(self, col, y):
        return col*10
ApplyToCols(transformer()).fit_transform(df)
```
that gives
```
      a   b
0   NaN  10
1   NaN  20
2   NaN  30
3  10.0  40
```

**Comment by emassoulie** [comment]
I'll check this one out as soon as https://github.com/skrub-data/skrub/issues/1820 is merged, to avoid conflicts with the refactoring of SingleColumnTransformer!

**Comment by rcap107** [comment]
Closing this as the main point was addressed in #1951 and the rest of the discussion should be continued in #1956
